### PR TITLE
feat: introduce app instance factory to allow extending app instantiation logic 

### PR DESCRIPTION
**Pull request by BeeHiveJava**
## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code compiles without warnings (code quality chek)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/net-daemon/docs

### CODE REVIEW

File: src/AppModel/NetDaemon.AppModel/Internal/Instance/AppInstanceHelper.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+namespace NetDaemon.AppModel.Internal.Resolver;
+
+internal static class AppInstanceHelper
+{
+    public static string GetAppId(Type type)
+    {
+        var attribute = type.GetCustomAttribute<NetDaemonAppAttribute>();
+        if (attribute is null)
+        {
+            throw new InvalidOperationException($"{type} has no {nameof(NetDaemonAppAttribute)}");
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Internal/Instance/AppInstanceHelper.cs:12]: If it does notbhave the attribite you can still fallback to the type name

File: src/AppModel/NetDaemon.AppModel/Internal/Instance/IAppInstance.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿namespace NetDaemon.AppModel.Internal.Resolver;
+
+internal interface IAppInstance
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Internal/Instance/IAppInstance.cs:3]: The name is a bit strange. This interfave does not represents an Instance but it creates instaces. The IAppInstanceResolver I think also does not have a correct name, it resolves apps, not instances of apps.

File: src/AppModel/NetDaemon.AppModel/Common/Extensions/ServiceCollectionExtension.cs [hunk 3]
```diff
@@ -95,13 +118,22 @@ internal static IServiceCollection AddAppModelIfNotExist(this IServiceCollection
 
     internal static IServiceCollection AddAppTypeResolverIfNotExist(this IServiceCollection services)
     {
-        // Check if we already registered
-        if (services.Any(n => n.ImplementationType == typeof(AppTypeResolver)))
-            return services;
+        // Check if already registered
+        if (services.All(descriptor => descriptor.ImplementationType != typeof(AppTypeResolver)))
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Common/Extensions/ServiceCollectionExtension.cs:122]: Personally I find the Any(.... == ) cleare than the All(.. !=  )  in this case

Comment (BeeHiveJava) [reply src/AppModel/NetDaemon.AppModel/Common/Extensions/ServiceCollectionExtension.cs:122]: But it's different. I inversed the logic here.

Comment (BeeHiveJava) [reply src/AppModel/NetDaemon.AppModel/Common/Extensions/ServiceCollectionExtension.cs:122]: The comment is outdated.

Comment (FrankBakkerNl) [reply src/AppModel/NetDaemon.AppModel/Common/Extensions/ServiceCollectionExtension.cs:122]: Ah but then I would prefer
if(!sercvies.Any(...))
I know Resharper.Rider will suggests the All here but I think it is strange to think of it that way. We do not do this because all registerd types are not of a specific type, but there is not any there yet that is of this type

File: src/AppModel/NetDaemon.AppModel/Internal/AppAssemblyProvider/SingleAppAssemblyProvider.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+using System.Reflection;
+
+namespace NetDaemon.AppModel.Internal.AppAssemblyProvider;
+
+internal class SingleAppAssemblyProvider : IAppAssemblyProvider
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Internal/AppAssemblyProvider/SingleAppAssemblyProvider.cs:5]: This is not just fo single apps (they dont even Use this provider I think) Maybe this should just be called AppAssemblyProvider

Comment (BeeHiveJava) [reply src/AppModel/NetDaemon.AppModel/Internal/AppAssemblyProvider/SingleAppAssemblyProvider.cs:5]: Then it will clash with the namespace. Maybe `DefaultAppAssemblyProvider`?

Comment (BeeHiveJava) [reply src/AppModel/NetDaemon.AppModel/Internal/AppAssemblyProvider/SingleAppAssemblyProvider.cs:5]: Or change the namespaces, I don't really like them anyway.

Comment (BeeHiveJava) [reply src/AppModel/NetDaemon.AppModel/Internal/AppAssemblyProvider/SingleAppAssemblyProvider.cs]: Renamed namespaces to be plural + renamed this class.

File: src/AppModel/NetDaemon.AppModel/Internal/AppFactories/AppFactoryHelper.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+﻿using System.Reflection;
+
+namespace NetDaemon.AppModel.Internal.AppFactories;
+
+internal static class AppFactoryHelper
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Internal/AppFactories/AppFactoryHelper.cs:5]: As there is just one place these are used now they could also be private memebers

Comment (BeeHiveJava) [reply src/AppModel/NetDaemon.AppModel/Internal/AppFactories/AppFactoryHelper.cs:5]: Used in `TypeAppFactory` and `FuncAppFactory`. One gets it based on a generic parameter, the other on a specifically passed `Type` parameter.

File: src/AppModel/NetDaemon.AppModel/Internal/AppFactories/TypeAppFactory.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+﻿namespace NetDaemon.AppModel.Internal.AppFactories;
+
+internal class TypeAppFactory : FuncAppFactory
+{
+    public TypeAppFactory(Type type, string? id = default, bool? focus = default) :
+        base(
+            CreateFactoryFunc(type),
+            id ?? AppFactoryHelper.GetAppId(type),
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Internal/AppFactories/TypeAppFactory.cs:8]: These are not needed hare as the base class also does that

Comment (BeeHiveJava) [reply src/AppModel/NetDaemon.AppModel/Internal/AppFactories/TypeAppFactory.cs:8]: Only the generic base class does this. I can move it to the non-generic base class, but then I'll need to work reflection magic to somehow get the returned `Type` from the `Func`.

Comment (BeeHiveJava) [reply src/AppModel/NetDaemon.AppModel/Internal/AppFactories/TypeAppFactory.cs:8]: Or pass the `Type` parameter in the constructor, but I don't like that because in theory it could differ from the `Type` in the `Func`.

Comment (BeeHiveJava) [reply src/AppModel/NetDaemon.AppModel/Internal/AppFactories/TypeAppFactory.cs:8]: Basically I only use these functions when I know the type is correct. That is two scenario's:
- You use the `FuncAppFactory` generic variant. In that case we enforce the generic parameter in the `Func`.
- You use the `TypeAppFactory`, in this case you specify the `Type` explicitly.

The third scenario is that you use the non-generic `FuncAppFactory`, but in that case you'll have to just pass the id and focus, because there's no reliable way to get it.

File: src/AppModel/NetDaemon.AppModel.Tests/Fixtures/Local/LocalAppWithId.cs [hunk 7]
```diff
@@ -1,6 +1,7 @@
 namespace LocalApps;
 
-[NetDaemonApp(Id = "SomeId")]
```
Comment (helto4real) [inline src/AppModel/NetDaemon.AppModel.Tests/Fixtures/Local/LocalAppWithId.cs:3]: [Why this change? Before we tested that user provided id was set byt the appmodel to the app. Maybe I am missing something :)
[Edit: Never mind you just use a const instead of hardcoded]

**Review by helto4real — APPROVED** [review]
I think you did an awesome job. The tests look really clean too!

File: src/AppModel/NetDaemon.AppModel/Common/IAppInstanceFactory.cs [hunk 8]
```diff
@@ -0,0 +1,6 @@
+﻿namespace NetDaemon.AppModel;
+
+public interface IAppInstanceFactory
```
[the remark this answers is not in the document]
Comment (BeeHiveJava) [reply src/AppModel/NetDaemon.AppModel/Common/IAppInstanceFactory.cs]: No, this should be removed.
